### ISSUE TITLE
Exit code not passed to calling process when using: #requires -Modules Pester

### ISSUE THREAD

**Issue by DaveKaretnyk** [pester/Pester#1021]
## 2. Describe Your Environment

Operating System, Pester version and PowerShell version:

```powershell
Pester version     : 4.3.1 C:\Program Files\WindowsPowerShell\Modules\Pester\4.3.1\Pester.psm1
PowerShell version : 5.1.16299.251
OS version         : Microsoft Windows NT 10.0.16299.0

Problem also seen on Windows 7 64-bit systems, and with earlier Pester versions.
```

## 3. Expected Behavior

When calling a PoSh script from a client program (e.g. another PoSh instance; a Windows cmd shell; or a C# application) the exit code from the PoSh script should be transmitted to the client program.

Example trivial script:
```powershell
#requires -Version 2.0
#requires -Modules Pester

Out-Host -InputObject ('64-bit OS?      {0}' -f [Environment]::Is64BitOperatingSystem)
Out-Host -InputObject ('64-bit process? {0}' -f [Environment]::Is64BitProcess)

Out-Host -InputObject ('Exit with code: {0}' -f 666777)

Exit 666777
```

Call the above from a PoSh or cmd shell:
```powershell
powershell.exe -NoProfile -ExecutionPolicy Bypass -File c:\tmp\pester\test_exit_pester.ps1
```
$LASTEXITCODE for PoSh should be 666777; %errorlevel% for cmd shell should be similar; i.e. any process executing this PoSh instance should get back this same error code.

This is the minimum simple example I could come up with to show the problem. Of course the real script I have does actually use Pester... ;-)

## 4.Current Behavior
See previous section; correct exit code not passed back to caller - always seems to be 0.

## 5. Possible Solution
Sorry, no ideas on what the problem is and how to solve it. The workaround which I finally figured out is to remove the #requires line involving Pester!?

## 6. Context
Usage of PoSh and Pester as part of integration testing on VMs.  The Pester based scripts are used to perform simple start/stop tests of applications. This works OK but the exit code is not properly sent back to the C# application that coordinates the testing - failed tests are not detected.

**Comment by nohwnd** [comment]
@DaveKaretnyk I am trying it, and I can observe the same behavior with many other modules. They don't seem to have anyting in common, but the behavior is reproducible, so maybe someone else will have a deeper look into this. 

Here is my repro: 

```powershell
$body = @'
#requires -Version 2.0
#requires -Modules ###name###

Exit 666777
'@

Get-Module -ListAvailable | ? name -NE isesteroids  | foreach { 

    ($body -replace "###name###", $_.Name) |  Set-Content $env:TEMP\f.ps1 
    powershell.exe -NoProfile -ExecutionPolicy Bypass -File $env:TEMP\f.ps1
    $_ | select ModuleType, Version, Name, @{n="exitCode"; e={$LASTEXITCODE}}
} 
```

As a workaround you could try using `[System.Environment]::Exit(666777)` which seems to work correctly.

**Comment by DaveKaretnyk** [comment]
@nohwnd: thanks! For investigating and for the workaround (works for me also). To be clear I understand, the workaround returns the exit value via the underlying .NET framework Exit rather than the PowerShell Exit?

Since it is a general PowerShell issue, is there somewhere that I can or should submit this?

**Comment by alx9r** [comment]
@DaveKaretnyk 
> Since it is a general PowerShell issue, is there somewhere that I can or should submit this?

I think I narrowed this down to script modules that dot-source.  I opened PowerShell/PowerShell#6615.

FWIW I have found exit codes from `powershell.exe` to often have surprising values.  That is somewhat mitigated by using `[System.Environment]::Exit()` as @nohwnd pointed out.  But even that technique can be frustrated, for example, when a `ParserError` occurs.  

It looks like there are a number of long-standing known issues related to exit codes:

* [stackoverflow.com](https://stackoverflow.com/a/15779295/1404637)
* [uservoice.com](https://windowsserver.uservoice.com/forums/301869-powershell/suggestions/11087790-powershell-exe-returns-a-passing-0-exit-code-whe)
